### PR TITLE
StudentFeedbackPageUiTest line 109 fails occasionally #3833

### LINKED ISSUE #3833: StudentFeedbackPageUiTest line 109 fails occasionally
This has been plaguing us recently, used to not happen at all so we need to investigate why.

Basically, Charlie and Danny doesn't get along and likes to switch places and cause the HTML verification to fail.

### PR DESCRIPTION
**Pull request by bamboo3250**
Fix #3833
@junhaoyap Could you take a look at this.

### CODE REVIEW

**Comment by junhaoyap** [comment]
Sure thing!

**Comment by junhaoyap** [comment]
I am consistently getting

screen shot 2015-07-06 at 3 21 50 pm

which seems to be correct since sorting by giver name should give Charlie before Danny, did you forget to change the HTML for this page?

**Comment by bamboo3250** [comment]
This doesn't happen to me when I ran the test before I created this PR but now I also got this. It's really annoying when we pass the test but not sure if it will 100% passed in the future or not. Please help me check this

**Comment by junhaoyap** [comment]
Yupp, I agree, we should keep working on these tests that can occasionally fail randomly and try to eliminate them!

I just ran the conflicting test 5 times and it passed every single time, I'm also currently running the full test suite so if it's green it's good to merge, will let you know by the end of today.

**Comment by junhaoyap** [comment]
Crap, I just got this when I ran the entire test suite...

screen shot 2015-07-06 at 4 53 16 pm

screen shot 2015-07-06 at 4 53 58 pm

And it's the only test that fails. Hmm, @kanghj this additional sort by giver doesn't seem to have solved the random failing test entirely, any insights?

**Comment by junhaoyap** [comment]
The weird thing is, I run the test alone 5 times again and it passes every single time, I guess it is still possible for this test to fail randomly when the entire test suite is ran?

**Comment by junhaoyap** [comment]
Let's still merge this in first, other than it maybe failing sometimes when the entire test suite is ran, at least it is a major improvement in that if it is run on its own it can pass for sure. We can keep an eye on this test after merging it in and if it continues to fail randomly on entire test suite runs (by merging it into master there will be more runs of the entire test suite) we can open another issue to fix that by building on this solution.

**Comment by kanghj** [comment]
It's actually because getNameForEmail doesn't handle teams. In the responses that are getting swapped, the giver type is Team. You will need to do a team comparison using <code>getTeamNameForEmail</code>, and then compare by the givers' names if the team is the same.

**Comment by junhaoyap** [comment]
Ok then let's do that @bamboo3250, Master Kang has spoken!

**Comment by bamboo3250** [comment]
Oh yeah, the function getNameForEmail is not completed. There is a TODO note there :+1:

**Comment by bamboo3250** [comment]
@junhaoyap Could you help me check this? 

Side Note:
The email of Danny was wrong. It was "SFResultsUiT.danny.e@gmail.tmt@gmail.tmt" and that resulted in a few changes in HTML files. 

Besides, I also found that some comparators can be replaced by other comparators. E.g compareByRecipientName and compareByRecipientGiverQuestion and so on.

**Comment by bamboo3250** [comment]
Oops, it still doesn't solve the problem. Continue to work on this.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 1]
```diff
@@ -1983,18 +1987,43 @@ public int compare(FeedbackResponseAttributes o1,
         }
     };
 
-    // Sorts by recipientName
+    // Sorts by recipientName > recipientEmail > giverName > giverEmail
     public final Comparator<FeedbackResponseAttributes> compareByRecipientName =
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: I'm not the reviewer, but I think you should rename this function

Comment (bamboo3250) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: maybe  compareByRecipientGiver?

**Comment by bamboo3250** [comment]
@junhaoyap Could you help me run the test and check this when you r free?

**Comment by junhaoyap** [comment]
Sure! So sorry I forgot! I wanted to do it yesterday but after being stuck in the traffic for 3-4 hours I forgot about this. Apologies! I'll get to it before lunch time :)

**Comment by bamboo3250** [comment]
Sure, no problem :+1:

**Comment by bamboo3250** [comment]
I ran all the tests and ran the StudentFeedbackResultsPageUITest a few time and it didn't happen but I still cannot really trust the test :joy:

**Comment by junhaoyap** [comment]
Too many random failures from this test eh, ok I'm looking at it and will start testing it after looking through the code, I'll let you know whether we have a better chance of trusting this thing soon :p

**Comment by junhaoyap** [comment]
Let's merge it in, it passes consistently! :+1:

**Comment by bamboo3250** [comment]
Okay, thanks :smile:

**Pull request merged**